src/ast/ast.c

```diff
@@ -206,14 +171,331 @@ AST *AST_Build(cypher_parse_result_t *parse_result) {
+	if(type == CYPHER_AST_RETURN) {
+		// Add referenced aliases for RETURN projections.
+		uint projectionCount = cypher_ast_return_nprojections(clause);
+		for(uint i = 0 ; i < projectionCount; i ++) {
+			_AST_MapProjectionAlias(ast, cypher_ast_return_get_projection(clause, i));
+		}
+		// Add referenced aliases for RETURN's ORDER BY entities.
+		const cypher_astnode_t *order_by = cypher_ast_return_get_order_by(clause);
+		if(order_by) _AST_MapOrderByReferences(ast, order_by);
+	} else if(type == CYPHER_AST_WITH) {
+		// Add referenced aliases for WITH projections.
+		uint projectionCount = cypher_ast_with_nprojections(clause);
+		for(uint i = 0 ; i < projectionCount; i ++) {
+			_AST_MapProjectionAlias(ast, cypher_ast_with_get_projection(clause, i));
+		}
+		// Add referenced aliases for WITH's ORDER BY entities.
+		const cypher_astnode_t *order_by = cypher_ast_with_get_order_by(clause);
+		if(order_by) _AST_MapOrderByReferences(ast, order_by);
+	} else if(type == CYPHER_AST_MATCH) {
+		// Add referenced aliases from MATCH clause - inline filtered and explicit WHERE filter.
+		_AST_MapMatchClauseReferences(ast, clause, include_all);
+	} else if(type == CYPHER_AST_CREATE) {
+		// Add referenced aliases for CREATE clause.
+		_AST_MapCreateClauseReferences(ast, clause, include_all);
+	} else if(type == CYPHER_AST_MERGE) {
+		// Add referenced aliases for MERGE clause - inline filtered and modified entities.
+		_AST_MapMergeClauseReference(ast, clause, include_all);
+	} else if(type == CYPHER_AST_SET) {
+		// Add referenced aliases for SET clause.
+		_AST_MapSetClauseReferences(ast, clause);
+	} else if(type == CYPHER_AST_DELETE) {
+		// Add referenced aliases for DELETE clause.
+		_AST_MapDeleteClauseReferences(ast, clause);
+	}
+}
+
+// Adds every reference entity in each clause.
+static void _AST_BuildReferenceMap(AST *ast, const cypher_astnode_t *project_clause) {
+	ast->referenced_entities = raxNew();
+
+	bool include_all = false;
+	if(project_clause) {
+		cypher_astnode_type_t type = cypher_astnode_type(project_clause);
+		assert(type == CYPHER_AST_WITH || type == CYPHER_AST_RETURN);
+		// If the projection clause is WITH/RETURN *, all user-defined aliases are referenced.
+		// Aliases can be introduced in MATCH, MERGE, CREATE, and UNWIND clauses.
+		if(type == CYPHER_AST_WITH) {
+			include_all = cypher_ast_with_has_include_existing(project_clause);
+		} else {
+			include_all = cypher_ast_return_has_include_existing(project_clause);
+		}
+
+		// If the projection is not a WITH/RETURN *, map its explicit references.
+		if(!include_all) _AST_MapReferredEntities(ast, project_clause);
+	}
+
+	// Check every clause in this AST.
+	uint clause_count = cypher_ast_query_nclauses(ast->root);
+	if(!clause_count) return;
+	for(uint i = 0; i < clause_count; i++) {
+		const cypher_astnode_t *clause = cypher_ast_query_get_clause(ast->root, i);
+		_ASTClause_BuildReferenceMap(ast, clause, include_all);
```

Comment:
No, the line you refer to is the sole caller of `_AST_MapReferredEntities` - it can't be triggered within `_ASTClause_BuildReferenceMap`. I just renamed it to `_AST_MapProjectionClause`, which is a bit more clear.